libraries/compute-utils/function-views/src/rich-function-view.ts

```diff
@@ -780,89 +856,189 @@ export class RichFunctionView extends FunctionView {
     if (prop.options['units']) t.addPostfix(prop.options['units']);
   }
 
+  private injectInputBaseValidation(t: DG.InputBase) {
+    const validationIndicator = ui.div('', {style: {display: 'flex'}});
+    t.addOptions(validationIndicator);
+    function setValidation(messages: ValidationResultBase | undefined) {
+      while (validationIndicator.firstChild && validationIndicator.removeChild(validationIndicator.firstChild));
+      const icon = getValidationIcon(messages);
+      if (icon)
+        validationIndicator.appendChild(icon);
```

Comment:
Don't see overcomplications with showing/hiding approach.
We can
1. create three icons in the very beginning (notif-ons, warnings, errors) 
2. show/hide them on each `setValidation` call (via CSS or directly)
3. replace the content of validation dialog right before the click

Custom inputs should properly override `addOptions` in this case - we will use DG.InputBase as an interface (which is not the best, but still).

Reply:
> Custom inputs should not override or implement addOptions in general 
Why? It is much more flexible and we should force input developers to do everything right. Where is no principal difference between implementing `addOptions` and `setValidation` methods. It is better to use InputBase as API, as we did it with its' `root`, `onInput` and other methods. It is how I see it strategically and it is important.

> It is more complicated, instead of having a fresh immutable tree, we need to mutate a part of it
If we keep the overall structure (input -> its' icon -> its' popover), mutations are fine. They do less job, they keep code simpler. This is not very important - I think, we can keep it and later I will show you how it could be done another way. In the end of the day, I may be wrong

Reply:
I see your solution is more flexible - but I don't like this flexibility. It is much better to force input developers to provide proper `addOptions` method and be control visual inside of it by the code of RFV.

In other words, if you have several stringInputs in the column (as single custom input) - no problem, provide a proper place for indication icons. It could be vertical div, a square in the corner, anything. Technically speaking, `FuncCallInput` class, from my point of view, will be united with InputBase in the end. This would provide an unified interface for all inputs.

I feel adding such a flexibility is a way to get crazy because of too much customization possibilities. And again, the code becomes a mess. If u r 100% sure that it is right way - OK then, I had to reassure that you see these suspicious places.